src/math/convolution/convolution.hpp

```diff
@@ -0,0 +1,202 @@
+      }
+      T rot = info.rate3[0];
+      for (usize s: rep(1, 1 << len)) {
+        usize offset = s << (h - len);
+        T rot2       = rot * rot;
+        T rot3       = rot2 * rot;
+        for (usize i: rep(0, p)) {
+          T a0                  = a[i + offset + 0 * p];
+          T a1                  = a[i + offset + 1 * p] * rot;
+          T a2                  = a[i + offset + 2 * p] * rot2;
+          T a3                  = a[i + offset + 3 * p] * rot3;
+          T a1na3imag           = (a1 - a3) * imag;
+          T a0a2                = a0 + a2;
+          T a1a3                = a1 + a3;
+          T a0na2               = a0 - a2;
+          a[i + offset + 0 * p] = a0a2 + a1a3;
+          a[i + offset + 1 * p] = a0a2 - a1a3;
+          a[i + offset + 2 * p] = a0na2 + a1na3imag;
+          a[i + offset + 3 * p] = a0na2 - a1na3imag;
+        }
+        rot *= info.rate3[__builtin_ctz(~s)];
+      }
+      len += 2;
+    }
+  }
+
+  template < typename T >
+  void butterfly_inv(std::vector< T > &a) {
+    usize n = a.size();
+    assert((n & (n - 1)) == 0);
+    usize h = __builtin_ctz(n);
+    ButterflyInfo< T > info;
+    usize len = h;
+    T iimag   = info.iroots[2];
+    while (len > 1) {
+      usize p = 1 << (h - len);
+      for (usize i: rep(0, p)) {
+        T a0         = a[i + 0 * p];
+        T a1         = a[i + 1 * p];
+        T a2         = a[i + 2 * p];
+        T a3         = a[i + 3 * p];
+        T a2na3iimag = (a2 - a3) * iimag;
+        T a0na1      = a0 - a1;
+        T a0a1       = a0 + a1;
+        T a2a3       = a2 + a3;
+        a[i + 0 * p] = a0a1 + a2a3;
+        a[i + 1 * p] = (a0na1 + a2na3iimag);
+        a[i + 2 * p] = (a0a1 - a2a3);
+        a[i + 3 * p] = (a0na1 - a2na3iimag);
+      }
+      T irot = info.irate3[0];
+      for (usize s: rep(1, 1 << (len - 2))) {
+        usize offset = s << (h - len + 2);
+        T irot2      = irot * irot;
+        T irot3      = irot2 * irot;
+        for (usize i: rep(0, p)) {
+          T a0                  = a[i + offset + 0 * p];
+          T a1                  = a[i + offset + 1 * p];
+          T a2                  = a[i + offset + 2 * p];
+          T a3                  = a[i + offset + 3 * p];
+          T a2na3iimag          = (a2 - a3) * iimag;
+          T a0na1               = a0 - a1;
+          T a0a1                = a0 + a1;
+          T a2a3                = a2 + a3;
+          a[i + offset + 0 * p] = a0a1 + a2a3;
+          a[i + offset + 1 * p] = (a0na1 + a2na3iimag) * irot;
+          a[i + offset + 2 * p] = (a0a1 - a2a3) * irot2;
+          a[i + offset + 3 * p] = (a0na1 - a2na3iimag) * irot3;
+        }
+        irot *= info.irate3[__builtin_ctz(~s)];
+      }
+      len -= 2;
+    }
+    if (len > 0) {
+      usize p = 1 << (h - 1);
+      for (usize i: rep(0, p)) {
+        T ajp = a[i] - a[i + p];
+        a[i] += a[i + p];
+        a[i + p] = ajp;
+      }
+    }
+  }
+} // namespace luz::internal
```

Comment:
namespace の終了の前のここ、こういう改行がなんかありがちなのでそうしましょう
```suggestion
  }

} // namespace luz::internal
```